### ISSUE TITLE
Bug: [no-generated-empty-object-type] False positive on a mapped type whose key set is a deferred conditional type

### ISSUE THREAD

**Issue by mnaoumov** [typescript-eslint/typescript-eslint#12853]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.ts&code=C4TwDgpgBA0hIGcA8AVAfFAvFFUIA9gIA7AEwSgEtiAzCAJygFUoB%2BKAa3gHsbmoAXFGIQAbgwDcAKCmhIUALIBDMJFKo8hEuSjcARgCsIAY2AZsAbygBtOCCrFY8ZOgC6Qu1AC%2B0qQHo-KAAlCDBueiJSIRQAC0oKOWh6CARuABtxBO4oAAMLLxyAGihgGOgIAFswUF1DE2AS8AgAOikCMIjG%2BRC6ZOJjCA0CIjIKfSNTc0UVNVQ0XwCoAEkG%2BOFuBsrq%2B25%2B6GoEYCViYEolSKhAFAISsoRoYAB3bLTqFK6IAFpjMuMOKCUKA96JRgCMBG18B0GsZdod-kIuCBeMEIL0SAMkFYlEJDsDiABzCRQPRCYgAVwqegY3imAHIlLTpIsAALABAfdr1Dn0ejhKAfKC04y0qgUYgbf6ceDFVI3e5NKDJJRpNL2Nb0kUAH0FelprXa4WhsOhCJ4fB6DHRgyxOOAeMJxNJFKpjC8dOF0iAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oHNEnFoAQ3yIAJvQC2NQrQ4AjAFaIydIqXRQB0DtEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Keys<T> = T extends infer U ? keyof U : never;

type Mapped<T extends object> = { [Key in Keys<T>]: Key };

// Reported: This type resolves to `{}`, the empty object type.
export type Reference<T extends object> = Mapped<T>;

// It is not empty once instantiated — these two lines type-check as written:
export const a: keyof Reference<{ a: string; b: number }> = 'a';
// @ts-expect-error - 'c' is not a key, so the type really is 'a' | 'b'.
export const c: keyof Reference<{ a: string; b: number }> = 'c';
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-generated-empty-object-type": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No report on line 6. `Mapped<T>` looks memberless only because `T` is not known yet — this is the
"types still awaiting type arguments" case the rule already means to skip. Every instantiation of it is a
real object type, never `{}`.

### Actual Result

Line 6, columns 43–52 reports:

> This type resolves to `{}`, the empty object type. This was likely not intentional.

Lines 9–11 show the type is not `{}`: `keyof Reference<{ a: string; b: number }>` accepts `'a'` and rejects
`'c'`, so it is exactly `'a' | 'b'`. `tsc --noEmit` is clean on that file — since an unused
`@ts-expect-error` is itself an error, its absence proves the rejection is real.

### Additional Info

The guard in `isEmptyObjectType` is `checker.isTypeAssignableTo(checker.getNumberType(), type)`, and a
deferred mapped type passes it. Querying the checker at the reported node gives:

```
isObjectType     : true
properties       : 0
indexInfos       : 0
callSignatures   : 0
ctorSignatures   : 0
numberAssignable : true    <-- the guard
```

So all five of the other conditions hold and the guard does not veto, and the rule reports.

Narrowing what actually triggers it: the `infer` is load-bearing, and the indexed access is not. These are
all clean on 8.70.0:

```ts
type A<T extends object> = { [K in keyof T]: K };
type A2<T extends object> = A<T>[keyof T];              // clean — homomorphic

type B<T extends string> = Record<T, unknown>;          // clean — the case the source comment names

type CKeys<T> = [T] extends [never] ? never : keyof T;
type C<T extends object> = { [K in CKeys<T>]: K };      // clean — conditional, but no `infer`

type DKeys<T> = T extends unknown ? keyof T : never;
type D<T extends object> = { [K in DKeys<T>]: K };      // clean — distributive, no `infer`
```

Real-world case: this shows up through `type-fest`'s `ConditionalKeys`, which resolves through nested
deferred conditionals. A "map, then index" key filter written on top of it —

```ts
type MethodKeysMap<T extends object> = {
  [Key in FunctionKeys<T>]: [Exclude<keyof NonNullable<T[Key]>, keyof GenericFunction>] extends [never] ? Key : never;
};
type MethodKeys<T extends object> = MethodKeysMap<T>[FunctionKeys<T>];
```

— goes red on the `MethodKeysMap<T>` reference, even though its `expectTypeOf` tests assert the instantiated
type is the correct key union. Since the rule is in `strictTypeChecked`, this reaches anything using that
idiom on the next dependency bump.

Local versions, which match the playground: eslint 10.10.0, typescript 6.0.3, typescript-eslint 8.70.0.
